### PR TITLE
Cleanup .trash WAL files

### PR DESCRIPTION
**Pull request by maysamyabandeh**
By rate limiting WAL deletion introduced in https://github.com/facebook/rocksdb/pull/5116 we could end up with WAL files with appendix ".trash". We should cleanup such files on db open similarly to other trash files.
Leave adding unit test as a TODO.

### CODE REVIEW

File: db/db_impl/db_impl_open.cc [hunk 1]
```diff
@@ -129,6 +129,7 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
   for (size_t i = 0; i < result.db_paths.size(); i++) {
     DeleteScheduler::CleanupDirectory(result.env, sfm, result.db_paths[i].path);
   }
+  DeleteScheduler::CleanupDirectory(result.env, sfm, std::string(result.wal_dir));
```
Comment (siying) [inline db/db_impl/db_impl_open.cc]: If user leaves options.wal_dir to be empty, to we end up with adding the same trash files from DB path to the cleaning up queue twice?

Comment (maysamyabandeh) [reply db/db_impl/db_impl_open.cc]: Fix that. Adding unit test is a good idea but it turns out to take more time. Will leave it for a separate PR.

File: db/db_impl/db_impl_open.cc [hunk 2]
```diff
@@ -129,6 +129,10 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
   for (size_t i = 0; i < result.db_paths.size(); i++) {
     DeleteScheduler::CleanupDirectory(result.env, sfm, result.db_paths[i].path);
   }
+  if (!result.wal_dir.empty()) {
```
Comment (anand1976) [inline db/db_impl/db_impl_open.cc]: Actually, ```wal_dir``` will not be empty as it is defaulted to dbname in line 95 if the user didn't set it. So maybe just compare it to dbname.

**Comment by anand1976** [comment]
The latest change LGTM. Since this is needed to unblock MyRocks, the unit test can be a separate PR.

File: db/db_impl/db_impl_open.cc [hunk 3]
```diff
@@ -92,9 +92,11 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
     result.wal_recovery_mode = WALRecoveryMode::kTolerateCorruptedTailRecords;
   }
 
+  bool separate_wal_dir = true;
   if (result.wal_dir.empty()) {
     // Use dbname as default
     result.wal_dir = dbname;
+    separate_wal_dir = false;
```
Comment (siying) [inline db/db_impl/db_impl_open.cc]: It's also possible that user provides a options.wal_dir which is the same as dbname. There might also be added twice, and I don't know the consequences for that.

Comment (maysamyabandeh) [reply db/db_impl/db_impl_open.cc]: I would compare the wal_dir with db_path but we would need a whole logic to skip extra slash, etc.

Comment (maysamyabandeh) [reply db/db_impl/db_impl_open.cc]: Taking look at the code it seems fine to double delete a file but given that guarantees of the internal APIs is not cleanly specified it is hard to be confident.

Comment (siying) [reply db/db_impl/db_impl_open.cc]: Can we at least still set separate_wal_dir if `result.wal_dir == dbname`? I know the byte-wise equality is not sufficient, but should be better than not doing it.

Btw, what's the behavior of double deleting?

Comment (maysamyabandeh) [reply db/db_impl/db_impl_open.cc]: My comment above says that it is hard to be confident about double deleting.
Sure. byte-wise comparison sounds good.

**Comment by siying** [comment]
Do we still need it?

**Comment by maysamyabandeh** [comment]
Closing in favor of https://github.com/facebook/rocksdb/pull/5520

**Pull request closed without merging**